### PR TITLE
Use `ParamSpec` to precisely annotate `@st.composite` and `st.functions()`

### PR DESCRIPTION
**Pull request by Zac-HD**
Being a follow-up to https://github.com/HypothesisWorks/hypothesis/pull/3219, now that [Mypy support for `Concatenate`](https://mypy-lang.blogspot.com/2022/04/mypy-0950-released.html) has been out for long enough for most users to pick up - since 0.950, April 2022.

(We explictly *don't* promise any degree of compatibility with anything but a best-effort attempt to support the latest state of the static-type-checking ecosystem, but giving people more than a few days to update like this seems reasonable to me.  The whole *point* of this is to provide a better user experience, after all!)

### CODE REVIEW

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py [hunk 1]
```diff
@@ -1931,52 +1964,114 @@ def emails() -> SearchStrategy[str]:
...
+else:  # pragma: no cover
+
+    @defines_strategy()
+    def functions(
+        *,
+        like: Callable[..., Any] = lambda: None,
+        returns: Union[SearchStrategy[Any], InferType] = infer,
+        pure: bool = False,
+    ) -> SearchStrategy[Callable[..., Any]]:
+        """functions(*, like=lambda: None, returns=..., pure=False)
```
Comment (sobolevn) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:2052]: Can we reuse `__doc__` somehow? 🤔

File: hypothesis-python/src/hypothesis/core.py [hunk 2]
```diff
@@ -1001,7 +1001,7 @@ def as_param(name, kind, defaults):
 def given(
     *_given_arguments: Union[SearchStrategy[Any], InferType],
 ) -> Callable[
-    [Callable[..., Union[None, Coroutine[Any, Any, None]]]], Callable[..., None]
+    [Callable[..., Optional[Coroutine[Any, Any, None]]]], Callable[..., None]
```
[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/core.py:1004]: Alas, yes, it's a partial-bind of *only* positional-or-keyword xor keyword-only parameters; and you can't pass strategies as positional args if there are any pos-only parameters; and positional args are bound from the rightmost pos-or-kw parameter (bizzare semantics, but we're stuck with it now).

I don't even see a more-precise overload we could extract, the closest would be "if you pass all the kwargs the wrapped function accepts, then the new function has no parameters", but even that breaks with `**kwargs`.

File: hypothesis-python/src/hypothesis/internal/filtering.py [hunk 3]
```diff
@@ -61,7 +61,7 @@ class ConstructivePredicate(NamedTuple):
     predicate: Optional[Predicate]
 
     @classmethod
-    def unchanged(cls, predicate):
+    def unchanged(cls, predicate: Predicate) -> "ConstructivePredicate":
```
[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/filtering.py:64]: This is a private/internal-only class and so should never be subclassed; I'll decorate it with `@final` to express that.

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py (hunk 1 above)
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/strategies/_internal/core.py:2052]: Unfortunately I couldn't see a way to do that which worked for docs-on-hover in IDEs; that's usually based on purely static information 😢

**Pull request merged**